File: Python/ByRound/2096/2096_G_Wonderful_Guessing_Game.py
# Problem: CF 2096 G - Wonderful Guessing Game
# https://codeforces.com/contest/2096/problem/G

"""
Algorithm: 
This solution uses a binary encoding strategy with base-3 representation to solve the adaptive guessing game.
The key idea is to:
1. Use base-3 digits (0, 1, 2) to encode each number from 1 to n
2. For each query, split the numbers into three categories:
   - 0: Not in the current query array (represented as 0 in digit)
   - 1: In the first half of query array (represented as 1 in digit) 
   - 2: In the second half of query array (represented as 2 in digit)
3. Since one query is ignored, we ensure we can recover the number even after one ignored query

Time Complexity: O(log_3(n) * n) per test case where we generate base-3 codes and do decoding
Space Complexity: O(log_3(n) * n) to store the base-3 representations and pair mappings
"""

import sys


def build_pairs(q):
    """return list of (c, partner(c)) for all words of length q
    with digit sum % 3 == 0, c < partner(c).  also caches the result."""
    total = 3**q

    sum_mod = [0] * total
    for i in range(1, total):
        sum_mod[i] = (sum_mod[i // 3] + (i % 3)) % 3

    pairs = []
    for c in range(total):
        if sum_mod[c] != 0:
            continue
        if c == 0:
            continue

        p = 0
        tmp = c
        power = 1
        for _ in range(q):
            d = tmp % 3
            if d == 1:
                d = 2
            elif d == 2:
                d = 1
            p += d * power
            tmp //= 3
            power *= 3
        if c < p:
            pairs.append((c, p))
    return pairs


def decode_word(x, q):
    """return list of q digits (0,1,2) of the base‑3 representation of x."""
    digits = [0] * q
    for i in range(q):
        digits[i] = x % 3
        x //= 3
    return digits


def solve():
    input_data = sys.stdin.readline
    t_line = input_data()
    while t_line.strip() == "":
        t_line = input_data()
    t = int(t_line)

    pair_cache = {}

    out = sys.stdout
    for _ in range(t):

        line = input_data()
        while line.strip() == "":
            line = input_data()
        n = int(line)

        q = 1
        while 3 ** (q - 1) < n:
            q += 1

        if q not in pair_cache:
            pair_cache[q] = build_pairs(q)
        pairs = pair_cache[q]

        if n % 2 == 0:
            need = n // 2
            use_zero = False
        else:
            need = (n - 1) // 2
            use_zero = True

        code = []
        if use_zero:
            code.append(0)
        for i in range(need):
            c, p = pairs[i]
            code.append(c)
            code.append(p)

        digits = [decode_word(w, q) for w in code]

        print(q)
        out.flush()

        for pos in range(q):
            left = []
            right = []
            for idx, d in enumerate(digits):
                val = d[pos]
                if val == 1:
                    left.append(idx + 1)
                elif val == 2:
                    right.append(idx + 1)
            arr = left + right
            print(len(arr), *arr)
            out.flush()

        s = input_data().strip()

        answer = None
        for idx, d in enumerate(digits):
            ok = True
            for i, ch in enumerate(s):
                if ch == "?":
                    continue
                digit = d[i]
                if ch == "L" and digit != 1:
                    ok = False
                    break
                if ch == "R" and digit != 2:
                    ok = False
                    break
                if ch == "N" and digit != 0:
                    ok = False
                    break
            if ok:
                answer = idx + 1
                break

        if answer is None:
            answer = 1
        print(answer)
        out.flush()


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/2096/2096_G_Wonderful_Guessing_Game.cpp
// Problem: CF 2096 G - Wonderful Guessing Game
// https://codeforces.com/contest/2096/problem/G

/*
G. Wonderful Guessing Game
Time Complexity: O(n log n) per test case
Space Complexity: O(n) 

This problem uses a divide-and-conquer strategy to construct a set of queries that can determine 
the hidden number in the minimum possible number of queries, f(n). The strategy is based on 
recursive partitioning where each level divides the search space into three parts:
- Left segment (marked as 1)
- Middle segment (not part of any query - marked as 0)
- Right segment (marked as 2)

The algorithm builds a ternary tree-like structure and for each level, assigns elements to 
different segments based on how they should be divided. Then, it constructs k-element arrays 
with even length that split the numbers into left and right parts in order to determine
the final answer.

Key techniques:
- Ternary partitioning using recursive calls
- Encoding query structure via array markings
- Adaptive querying with precomputed responses
*/

#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

#define F first
#define S second
#define all(x) x.begin(), x.end()
#define pb push_back
#define FIO                                                                    \
  ios_base::sync_with_stdio(false);                                            \
  cin.tie(0);                                                                  \
  cout.tie(0)
using namespace std;
typedef long long ll;
typedef pair<ll, ll> pii;
const int N = 2e5 + 7;
int a[20][N];

// Recursive function to mark the divisions in an array.
void f(int l, int r, int d) {
  if (l >= r)
    return;
  int m = (r - l + 2) / 3; // Calculate the size of left/right parts
  for (int i = l; i < l + m; i++)
    a[d][i] = 1; // Mark elements in left part as 1 (Left query)
  for (int i = r - m + 1; i <= r; i++)
    a[d][i] = 2; // Mark elements in right part as 2 (Right query)
  f(l, l + m - 1, d + 1);      // Recursively process left segment
  f(l + m, r - m, d + 1);      // Recursively process middle segment
  f(r - m + 1, r, d + 1);      // Recursively process right segment
}

void task() {
  int n;
  cin >> n;
  for (int i = 0; i < 20; i++)
    fill(a[i], a[i] + n, 0);
  int q = 0, qq = 1;
  while (qq < n)
    qq *= 3, q++; // Determine number of levels needed
  f(0, n - 1, 0);   // Start the recursive marking process

  // Sum up the markings and apply modulo 3 logic to determine final assignments
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < q; j++)
      a[q][i] += a[j][i]; // Accumulate marking levels
    a[q][i] = (3 - a[q][i] % 3) % 3; // Final mapping to 0/1/2
  }

  cout << q + 1 << endl;

  vector<int> ok(n, 1);

  // Output the constructed queries for each level of recursion
  for (int i = 0; i <= q; i++) {
    vector<int> u, v;
    for (int j = 0; j < n; j++) {
      if (a[i][j] == 1)
        u.pb(j + 1); // Collect elements belonging to left query part
      if (a[i][j] == 2)
        v.pb(j + 1); // Collect elements belonging to right query part
    }
    cout << u.size() + v.size() << " ";
    for (int x : u)
      cout << x << " ";
    for (int x : v)
      cout << x << " ";
    cout << endl;
  }

  string s;
  cin >> s; // Read the responses from Alice

  // Process responses to filter out impossible numbers
  for (int i = 0; i <= q; i++) {
    if (s[i] == 'L') {
      for (int j = 0; j < n; j++)
        if (a[i][j] != 1) // If a number doesn't belong to left part, it's invalid
          ok[j] = 0;
    } else if (s[i] == 'R') {
      for (int j = 0; j < n; j++)
        if (a[i][j] != 2) // If a number doesn't belong to right part, it's invalid
          ok[j] = 0;
    } else if (s[i] == 'N') {
      for (int j = 0; j < n; j++)
        if (a[i][j]) // If a number is in any query part, it's invalid
          ok[j] = 0;
    }
  }

  // Output the final determined number
  for (int i = 0; i < n; i++)
    if (ok[i]) {
      cout << i + 1 << endl;
      break;
    }
}

int main() {
  FIO;
  int tt;
  cin >> tt;
  while (tt--)
    task();
  return 0;
}


// https://github.com/VaHiX/codeForces/